### PR TITLE
Use pkttyagent to request user passwords if running without GUI

### PR DESCRIPTION
**Pull request by hughsie**
This change will allow it to use pkcon over remote-shells (ssh) or
to use it witout a running GUI desktop environment in the background.

Should fix https://github.com/fwupd/fwupd/issues/2429

### CODE REVIEW

File: src/fu-util.c [hunk 1]
```diff
@@ -2845,6 +2846,9 @@ main (int argc, char *argv[])
 	if (no_history)
 		priv->flags |= FWUPD_INSTALL_FLAG_NO_HISTORY;
 
+	/* start polkit tty agent to listen for password requests */
+	fu_polkit_agent_open ();
```
Comment (superm1) [inline src/fu-util.c:2850]: does this need to run on every run of the app?

Comment (superm1) [reply src/fu-util.c:2850]: My thought being for example it wouldn't run in non-interactive mode

Comment (superm1) [inline src/fu-util.c:2850]: the return code seems to be int, shouldn't GError be passed in and the return code be boolean instead (maybe with a local error)?

Comment (hughsie) [reply src/fu-util.c:2850]: Yes, it gets killed after the command quits. It's exactly the same logic as pkcon has been using for ~ 5 years.

Comment (hughsie) [reply src/fu-util.c:2850]: Yes, good idea.

File: src/fu-polkit-agent.c [hunk 2]
```diff
@@ -0,0 +1,219 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2011 Lennart Poettering <lennart@poettering.net>
+ * Copyright (C) 2012 Matthias Klumpp <matthias@tenstral.net>
```
Comment (superm1) [inline src/fu-polkit-agent.c:4]: I guess this was lifted from systemd?  It feels like systemd style code.  If it was new code I guess a few things would have been differently like not using variable arguments.

File: src/fu-polkit-agent.c [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+#include <fcntl.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <sys/poll.h>
+#include <sys/wait.h>
+#include <glib.h>
+
+#include "fu-polkit-agent.h"
+
+#define POLKIT_TTY_AGENT_BINARY_PATH "/usr/bin/pkttyagent"
```
Comment (superm1) [inline src/fu-polkit-agent.c:29]: Probably shouldn't hardcode this path, we have infrastructure to lookup paths as configured by distros.

File: src/fu-polkit-agent.c (hunk 2 above)
Comment (hughsie) [reply src/fu-polkit-agent.c:4]: > I guess this was lifted from systemd

Yes, a long time ago (LGPLv2+ also) -- we could clean up some stuff, but I'm hesitant to change the low level UNIX bits as in all honesty I don't understand all the nuances mentioned.

File: src/fu-polkit-agent.c (hunk 3 above)
Comment (hughsie) [reply src/fu-polkit-agent.c:29]: Fixed.

File: src/fu-util.c (hunk 1 above)
Comment (hughsie) [reply src/fu-util.c:2850]: Done,a few things cleaned up.

**Comment by superm1** [comment]
So it's a bit awkward using this since there is a fighting for stdout.  

```
$ fwupdmgr enable-remote lvfs-testing
╔══════════════════════════════════════════════════════════════════════════════╗
║ The LVFS is a free service that operates as an independent legal entity and  ║
║ has no connection with Ubuntu. Your distributor may not have verified any    ║
║ of the firmware updates for compatibility with your system or connected      ║
║ devices. All firmware is provided only by the original equipment             ║
║ manufacturer.                                                                ║
║                                                                              ║
║ This remote contains firmware which is not embargoed, but is still being     ║
║ tested by the hardware vendor. You should ensure you have a way to manually  ║
║ downgrade the firmware if the firmware update fails.                         ║
║                                                                              ║
║ Enabling this functionality is done at your own risk, which means you have   ║
║ to contact your original equipment manufacturer regarding any problems       ║
║ caused by these updates. Only problems with the update process itself        ║
║ should be filed at https://bugs.launchpad.net/ubuntu/.                       ║
║                                                                              ║
╚══════════════════════════════════════════════════════════════════════════════╝

Agree and enable the remote? [Y|n]: y
Authenticating…          [-                                      ]==== AUTHENTICATING FOR org.freedesktop.fwupd.modify-remote ===
Authentication is required to modify a configured remote used for firmware updates
Authenticating as: supermario,,, (supermario)
Authenticating…          [                                     \ ]
Authenticating…          [                                      \]==== AUTHENTICATION COMPLETE ===
Authenticating…          [***************************************]
Successfully enabled remote
```

File: src/fu-util.c [hunk 4]
```diff
@@ -2845,6 +2847,12 @@ main (int argc, char *argv[])
 	if (no_history)
 		priv->flags |= FWUPD_INSTALL_FLAG_NO_HISTORY;
 
+	/* start polkit tty agent to listen for password requests */
```
Comment (superm1) [inline src/fu-util.c:2850]: don't you also want to check `is_interactive`?

Comment (hughsie) [reply src/fu-util.c:2850]: Yes! Fixed.

**Comment by hughsie** [comment]
> fighting for stdout

How's that new version working? ^^

**Review by superm1 — APPROVED** [review]
Much better.  branch looks good to me now.

**Pull request merged**